numpy/array_api/linalg.py

```diff
@@ -379,7 +379,18 @@ def trace(x: Array, /, *, offset: int = 0) -> Array:
 def vecdot(x1: Array, x2: Array, /, *, axis: int = -1) -> Array:
     if x1.dtype not in _numeric_dtypes or x2.dtype not in _numeric_dtypes:
         raise TypeError('Only numeric dtypes are allowed in vecdot')
-    return tensordot(x1, x2, axes=((axis,), (axis,)))
+    ndim = max(x1.ndim, x2.ndim)
+    x1_shape = (1,)*(ndim - x1.ndim) + tuple(x1.shape)
+    x2_shape = (1,)*(ndim - x2.ndim) + tuple(x2.shape)
```

Comment:
This will go away with the change.

Reply:
Maybe this can be simplified, but without this, the shape checking is not correct (because size 1 dimensions broadcast). 

Also, the broadcasting is necessary because, assuming I am understanding the idea in the spec correctly, nonnegative axes should refer to the broadcast dimensions.

Reply:
You can move the checking to later and use `-1` as axis.  That probably gives you better errors for bad axis values anyway.

Reply:
I'm sorry, I'm not following your suggestion here.

Reply:
I am saying that you can do the check after the broadcast+moveaxis at the point where the axis is properly normalized to `-1`.

Reply:
I took the liberty to do that change, unless you disagree will merge next time around (or earlier with feedback).